src/cargo/core/compiler/job_queue.rs

```diff
@@ -284,129 +616,33 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
-                    }
-                    Message::BuildPlanMsg(module_name, cmd, filenames) => {
-                        plan.update(&module_name, &cmd, &filenames)?;
-                    }
-                    Message::Stdout(out) => {
-                        cx.bcx.config.shell().stdout_println(out);
-                    }
-                    Message::Stderr(err) => {
-                        let mut shell = cx.bcx.config.shell();
-                        shell.print_ansi(err.as_bytes())?;
-                        shell.err().write_all(b"\n")?;
-                    }
-                    Message::FixDiagnostic(msg) => {
-                        print.print(&msg)?;
-                    }
-                    Message::Finish(id, artifact, result) => {
-                        let unit = match artifact {
-                            // If `id` has completely finished we remove it
-                            // from the `active` map ...
-                            Artifact::All => {
-                                info!("end: {:?}", id);
-                                finished += 1;
-                                self.active.remove(&id).unwrap()
-                            }
-                            // ... otherwise if it hasn't finished we leave it
-                            // in there as we'll get another `Finish` later on.
-                            Artifact::Metadata => {
-                                info!("end (meta): {:?}", id);
-                                self.active[&id]
-                            }
-                        };
-                        info!("end ({:?}): {:?}", unit, result);
-                        match result {
-                            Ok(()) => self.finish(id, &unit, artifact, cx)?,
-                            Err(e) => {
-                                let msg = "The following warnings were emitted during compilation:";
-                                self.emit_warnings(Some(msg), &unit, cx)?;
-
-                                if !self.active.is_empty() {
-                                    error = Some(anyhow::format_err!("build failed"));
-                                    handle_error(&e, &mut *cx.bcx.config.shell());
-                                    cx.bcx.config.shell().warn(
-                                        "build failed, waiting for other \
-                                         jobs to finish...",
-                                    )?;
-                                } else {
-                                    error = Some(e);
-                                }
-                            }
-                        }
-                    }
-                    Message::Token(acquired_token) => {
-                        tokens.push(
-                            acquired_token.chain_err(|| "failed to acquire jobserver token")?,
-                        );
-                    }
+            self.tokens.truncate(self.active.len() - 1);
```

Comment:
Could this get moved right before the actual blocking? That way we wouldn't accidentally throw away tokens while we're processing a bunch of messages.